{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "5_a5lMllIjX9"
   },
   "source": [
    "# Phase separation prediction\n",
    "\n",
    "This Colab notebook enables prediction of IDR transfer free energies and saturation concentrations from sequence.\n",
    "- The models have been trained on CALVADOS 2 slab simulation data.\n",
    "- Conditions are fixed to T=293 K and I=150 mM.\n",
    "\n",
    "<b>How to cite this notebook:</b>\n",
    "- S. von Bülow, G. Tesei & K. Lindorff-Larsen, __Prediction of phase separation propensities of disordered proteins from sequence__ _bioRxiv_ 2024\n",
    "\n",
    "Further references:\n",
    "- Use of $\\nu_\\mathrm{SVR}$: <br>\n",
    "G. Tesei, A. I. Trolle, N. Jonsson, J. Betz, F. Pesce, K. E. Johansson, K. Lindorff-Larsen, __Conformational ensembles of the human intrinsically disordered proteome__ _Nature_ 2024 626, 897–904 DOI: https://doi.org/10.1038/s41586-023-07004-5\n",
    "- CALVADOS 2 model: <br>\n",
    "G. Tesei and K. Lindorff-Larsen, __Improved predictions of phase behaviour of intrinsically disordered proteins by tuning the interaction range [version 2; peer review: 2 approved]__ _Open Research Europe_ 2023 2(94) DOI: https://doi.org/10.12688/openreseurope.14967.2\n",
    "\n",
    "Author: Sören von Bülow (soren.bulow@bio.ku.dk)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "1CsZ6Y9fywzl"
   },
   "outputs": [],
   "source": [
    "#@title <b>Preliminary operations</b>\n",
    "\n",
    "import os\n",
    "import warnings\n",
    "warnings.simplefilter(\"ignore\")\n",
    "\n",
    "print('Setting up the environment...')\n",
    "\n",
    "!rm -r sample_data &> dump\n",
    "!rm sequence.* &> dump\n",
    "!rm svr_model_nu* &> dump\n",
    "!rm residues* &> dump\n",
    "!rm example* &> dump\n",
    "!rm predictor.* &> dump\n",
    "!rm *joblib &> dump\n",
    "\n",
    "github_folder = 'https://raw.githubusercontent.com/KULL-Centre/_2024_buelow_PSpred/main'\n",
    "\n",
    "print(f'Downloading files from {github_folder}')\n",
    "\n",
    "os.system(f'wget {github_folder}/scripts_colab/sequence.py')\n",
    "os.system(f'wget {github_folder}/scripts_colab/predictor.py')\n",
    "os.system(f'wget {github_folder}/data/residues.csv')\n",
    "os.system(f'wget {github_folder}/data/example.fasta')\n",
    "os.system(f'wget -O model_dG.joblib {github_folder}/models/idrome90/mlp/dG/model.joblib')\n",
    "os.system(f'wget -O model_logcdil_mgml.joblib {github_folder}/models/idrome90/mlp/logcdil_mgml/model.joblib')\n",
    "os.system(f'wget -O svr_model_nu.joblib {github_folder}/models/svr_model_nu.joblib')\n",
    "\n",
    "!pip install 'scikit-learn==1.3' MDAnalysis biopython numba &> dump\n",
    "\n",
    "import joblib\n",
    "import sklearn\n",
    "import pandas as pd\n",
    "\n",
    "import sequence\n",
    "from predictor import *\n",
    "\n",
    "import numpy as np\n",
    "import MDAnalysis as mda\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from tqdm import tqdm\n",
    "from google.colab import files\n",
    "\n",
    "ncrossval = 50\n",
    "print('Environment set up.')\n",
    "print('Loading models...')\n",
    "\n",
    "residues = pd.read_csv('residues.csv').set_index('one')\n",
    "nu_file = 'svr_model_nu.joblib'\n",
    "\n",
    "features = ['mean_lambda', 'faro', 'shd', 'ncpr', 'fcr', 'scd', 'ah_ij','nu_svr']\n",
    "\n",
    "features_clean = {\n",
    "    'mean_lambda' : 'lambda',\n",
    "    'faro' : 'f(aromatics)',\n",
    "    'shd' : 'SHD',\n",
    "    'ncpr' : 'NCPR',\n",
    "    'fcr' : 'FCR',\n",
    "    'scd' : 'SCD',\n",
    "    'ah_ij' : 'LJ pairs',\n",
    "    'nu_svr' : 'nu(SVR)'\n",
    "}\n",
    "\n",
    "print('Input features are:')\n",
    "print('>>>>> '+ ', '.join([features_clean[fe] for fe in features]))\n",
    "\n",
    "!touch calvados.py\n",
    "\n",
    "models = {}\n",
    "models['dG'] = joblib.load(f'model_dG.joblib')\n",
    "models['logcdil_mgml'] = joblib.load(f'model_logcdil_mgml.joblib')\n",
    "\n",
    "mltype = 'mlp'\n",
    "alpha = 5\n",
    "layers = (10,10)\n",
    "\n",
    "targets = ['dG','logcdil_mgml']\n",
    "targets_clean = {\n",
    "    'dG' : 'Delta G',\n",
    "    'logcdil_mgml' : 'Saturation concentration',\n",
    "}\n",
    "\n",
    "print('Models loaded.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "0F7WtrIChg-F"
   },
   "outputs": [],
   "source": [
    "#@title <b>Predict single IDR sequence</font></b>\n",
    "\n",
    "try:\n",
    "    os.rmdir('sample_data')\n",
    "except:\n",
    "    pass\n",
    "\n",
    "NAME = \"LAF1\" #@param {type:\"string\"}\n",
    "SEQUENCE = \"MESNQSNNGGSGNAALNRGGRYVPPHLRGGDGGAAAAASAGGDDRRGGAGGGGYRRGGGNSGGGGGGGYDRGYNDNRDDRDNRGGSGGYGRDRNYEDRGYNGGGGGGGNRGYNNNRGGGGGGYNRQDRGDGGSSNFSRGGYNNRDEGSDNRGSGRSYNNDRRDNGGDGLEHHHHHH\" #@param {type:\"string\"}\n",
    "CHARGE_TERMINI = True # @param {type:'boolean'}\n",
    "TEMPERATURE = \"293 K (fixed)\" # @param ['293 K (fixed)']\n",
    "IONIC_STRENGTH = \"150 mM (fixed)\" # @param ['150 mM (fixed)']\n",
    "\n",
    "seq = SEQUENCE\n",
    "if \" \" in seq:\n",
    "    seq = ''.join(seq.split())\n",
    "    print('Blank character(s) found in the provided sequence. Sequence has been corrected, but check for integrity.')\n",
    "\n",
    "print('='*80)\n",
    "print(f'NAME: {NAME}')\n",
    "print(f'SEQUENCE: {seq}')\n",
    "\n",
    "seqfeats = sequence.SeqFeatures(seq,residues=residues,charge_termini=CHARGE_TERMINI)\n",
    "X = X_from_seq(seq,features,residues=residues,charge_termini=CHARGE_TERMINI,nu_file=nu_file)\n",
    "\n",
    "for target in targets:\n",
    "  print('-'*80)\n",
    "  ys = models[target].predict(X)#,models)\n",
    "  ys_m = np.mean(ys)\n",
    "\n",
    "  if target == 'dG':\n",
    "    output = ys_m\n",
    "    unit = 'kT'\n",
    "    lower = ys_m - 1\n",
    "    upper = ys_m + 1\n",
    "  elif target == 'logcdil_mgml':\n",
    "    output = np.exp(ys_m)\n",
    "    lower = np.exp(ys_m-0.82)\n",
    "    upper = np.exp(ys_m+0.82)\n",
    "    unit = 'mg/mL'\n",
    "\n",
    "  print(f'{targets_clean[target]:25s} = {output:5.1f} {unit:6s} ({lower:.1f} -- {upper:.1f} {unit})')\n",
    "  if target == 'logcdil_mgml':\n",
    "    output_uM = output / seqfeats.mw * 1e6\n",
    "    lower_uM = lower / seqfeats.mw * 1e6\n",
    "    upper_uM = upper / seqfeats.mw * 1e6\n",
    "    print(f'{\"\":25s} = {output_uM:5.1f} {\"uM\":6s} ({lower_uM:.1f} -- {upper_uM:.1f} {\"uM\"})')\n",
    "print('='*80)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "L9RekX9rtunn"
   },
   "outputs": [],
   "source": [
    "#@title <b>Run batch prediction</b>\n",
    "\n",
    "#@markdown File name\n",
    "FASTA_FILE = \"example.fasta\" #@param {type:\"string\"}\n",
    "CHARGE_TERMINI = True # @param {type:'boolean'}\n",
    "TEMPERATURE = \"293 K (fixed)\" # @param ['293 K (fixed)']\n",
    "IONIC_STRENGTH = \"150 uM (fixed)\" # @param ['150 uM (fixed)']\n",
    "\n",
    "if not os.path.isfile(FASTA_FILE):\n",
    "  print(f'Please upload file {FASTA_FILE}')\n",
    "  uploaded = files.upload()\n",
    "  if FASTA_FILE not in uploaded.keys():\n",
    "    raise NameError(f'Could not find file {FASTA_FILE}')\n",
    "\n",
    "records = sequence.read_fasta(FASTA_FILE)\n",
    "\n",
    "print('-'*80)\n",
    "print(f'FASTA FILE: {FASTA_FILE}')\n",
    "print(f'NUMBER OF SEQUENCES: {len(records)}')\n",
    "\n",
    "#@title <b>Run batch prediction</font></b>\n",
    "\n",
    "df_records = pd.DataFrame(dtype=object)\n",
    "\n",
    "for name, record in tqdm(records.items(),total=len(records)):\n",
    "  seq = str(record.seq)\n",
    "  df_records.loc[name,'Sequence'] = seq\n",
    "  seqfeats = sequence.SeqFeatures(seq,residues=residues,\n",
    "                                  charge_termini=CHARGE_TERMINI,nu_file=nu_file)\n",
    "  for feat in features:\n",
    "    df_records.loc[name,feat] = getattr(seqfeats,feat)\n",
    "  X = X_from_seq(seq,features,residues=residues,\n",
    "                 charge_termini=CHARGE_TERMINI,nu_file=nu_file)\n",
    "  for target in targets:\n",
    "    ys = models[target].predict(X)#,models)\n",
    "    ys_m = np.mean(ys)\n",
    "    if target == 'dG':\n",
    "      df_records.loc[name,'Delta G [kT]'] = ys_m\n",
    "    if target == 'logcdil_mgml':\n",
    "      cdil_mgml = np.exp(ys_m)\n",
    "      df_records.loc[name,'Saturation concentration [mg/mL]'] = cdil_mgml\n",
    "      cdil_uM = cdil_mgml / seqfeats.mw * 1e6\n",
    "      df_records.loc[name,'Saturation concentration [uM]'] = cdil_uM\n",
    "\n",
    "df_records.index.name = 'Name'\n",
    "df_records.to_csv('df_PSprediction.csv')\n",
    "\n",
    "print('\\n')\n",
    "print('='*114)\n",
    "print(f'{\"Name\":20s} {\"Sequence\":33s} {\"Delta G\":>10s} {\"Saturation\":>16s} {\"Saturation\":>16s}')\n",
    "print(f'{\"\":20s} {\"\":33s} {\"\":>10s} {\"concentration\":>16s} {\"concentration\":>16s}')\n",
    "print(f'{\"\":20s} {\"\":33s} {\"[kT]\":>10s} {\"[mg/mL]\":>16s} {\"[uM]\":>16s}')\n",
    "\n",
    "print('='*114)\n",
    "for key, val in df_records.iterrows():\n",
    "  if len(list(val[\"Sequence\"])) > 30:\n",
    "    seqpr = f'{val[\"Sequence\"][:30]:30s}...'\n",
    "  else:\n",
    "    seqpr = f'{val[\"Sequence\"][:30]:30s}'\n",
    "  print(f'{key:20s} {seqpr:33s} {val[\"Delta G [kT]\"]:10.1f} {val[\"Saturation concentration [mg/mL]\"]:16.1f} {val[\"Saturation concentration [uM]\"]:16.1f}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "bNX8tEIW29um"
   },
   "outputs": [],
   "source": [
    "#@title <b>Analysis</font></b>\n",
    "\n",
    "fcolor = plt.cm.summer\n",
    "\n",
    "fig, ax = plt.subplots(1,2,figsize=(9,4))\n",
    "\n",
    "for idx, target in enumerate(['Delta G [kT]','Saturation concentration [mg/mL]']):\n",
    "  axij = ax[idx]\n",
    "  axij.hist(df_records[target],bins=20,color=fcolor(0))\n",
    "\n",
    "  axij.set_xlabel(f'{target}')\n",
    "  axij.set_ylabel('Counts')\n",
    "  axij.grid(alpha=0.3)\n",
    "fig.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "-SCC7r8ltUTy"
   },
   "outputs": [],
   "source": [
    "#@title <b>Download dataframe</font></b>\n",
    "\n",
    "files.download('df_PSprediction.csv')"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
