{"cells":[{"cell_type":"markdown","metadata":{"id":"5_a5lMllIjX9"},"source":["# Phase separation prediction (TWIN2PIPSA version)\n","\n","This Colab notebook enables prediction of IDR transfer free energies and saturation concentrations from sequence.\n","- The models have been trained on CALVADOS 2 slab simulation data.\n","- Conditions are fixed to T=293 K and I=150 mM.\n","\n","<b>How to cite this notebook:</b>\n","- S. von Bülow, G. Tesei, F. K. Zaidi, T. Mittag, K. Lindorff-Larsen, __Prediction of phase-separation propensities of disordered proteins from sequence__ _Proceedings of the National Academy of Sciences_ 2025\n","\n","Further references:\n","- Use of $\\nu_\\mathrm{SVR}$: <br>\n","G. Tesei, A. I. Trolle, N. Jonsson, J. Betz, F. Pesce, K. E. Johansson, K. Lindorff-Larsen, __Conformational ensembles of the human intrinsically disordered proteome__ _Nature_ 2024 626, 897–904 DOI: https://doi.org/10.1038/s41586-023-07004-5\n","- CALVADOS 2 model: <br>\n","G. Tesei and K. Lindorff-Larsen, __Improved predictions of phase behaviour of intrinsically disordered proteins by tuning the interaction range [version 2; peer review: 2 approved]__ _Open Research Europe_ 2023 2(94) DOI: https://doi.org/10.12688/openreseurope.14967.2\n","\n","Author: Sören von Bülow (soren.bulow@bio.ku.dk)"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","id":"1CsZ6Y9fywzl"},"outputs":[],"source":["#@title <b>Preliminary operations</b>\n","\n","import os\n","import warnings\n","warnings.simplefilter(\"ignore\")\n","\n","print('Setting up the environment...')\n","\n","!rm -r sample_data &> dump\n","!rm sequence.* &> dump\n","!rm svr_model_nu* &> dump\n","!rm residues* &> dump\n","!rm example* &> dump\n","!rm predictor.* &> dump\n","!rm *joblib &> dump\n","\n","github_folder = 'https://raw.githubusercontent.com/KULL-Centre/_2024_buelow_PSpred/twin2pipsa'\n","\n","print(f'Downloading files from {github_folder}')\n","\n","os.system(f'wget {github_folder}/scripts_colab/sequence.py')\n","os.system(f'wget {github_folder}/scripts_colab/predictor.py')\n","os.system(f'wget {github_folder}/data/residues.csv')\n","os.system(f'wget {github_folder}/data/example.fasta')\n","os.system(f'wget -O model_dG.joblib {github_folder}/models/idrome90/mlp/dG/model.joblib')\n","os.system(f'wget -O model_logcdil_mgml.joblib {github_folder}/models/idrome90/mlp/logcdil_mgml/model.joblib')\n","os.system(f'wget -O svr_model_nu.joblib {github_folder}/models/svr_model_nu.joblib')\n","\n","os.system(f'wget {github_folder}/data/IDRome_DB_full.csv')\n","\n","!pip install 'scikit-learn==1.3' MDAnalysis biopython numba &> dump\n","\n","import joblib\n","import sklearn\n","import pandas as pd\n","\n","import sequence\n","from predictor import *\n","\n","import numpy as np\n","import MDAnalysis as mda\n","import matplotlib.pyplot as plt\n","\n","from tqdm import tqdm\n","from google.colab import files\n","\n","ncrossval = 50\n","print('Environment set up.')\n","print('Loading models...')\n","\n","colors = np.array([\n","    [1., 101., 165.],\n","    [220., 140., 46.], # [220., 175., 46.],\n","    [31., 107., 65.],\n","    [200., 80., 45.],\n","    [127., 0., 255.]\n","])\n","colors /= 255.\n","\n","residues = pd.read_csv('residues.csv').set_index('one')\n","nu_file = 'svr_model_nu.joblib'\n","\n","df_full = pd.read_csv('IDRome_DB_full.csv').set_index('id')\n","\n","features = ['mean_lambda', 'faro', 'shd', 'ncpr', 'fcr', 'scd', 'ah_ij','nu_svr']\n","\n","features_clean = {\n","    'mean_lambda' : 'lambda',\n","    'faro' : 'f(aromatics)',\n","    'shd' : 'SHD',\n","    'ncpr' : 'NCPR',\n","    'fcr' : 'FCR',\n","    'scd' : 'SCD',\n","    'ah_ij' : 'LJ pairs',\n","    'nu_svr' : 'nu(SVR)'\n","}\n","\n","print('Input features are:')\n","print('>>>>> '+ ', '.join([features_clean[fe] for fe in features]))\n","\n","!touch calvados.py\n","\n","models = {}\n","models['dG'] = joblib.load(f'model_dG.joblib')\n","models['logcdil_mgml'] = joblib.load(f'model_logcdil_mgml.joblib')\n","\n","mltype = 'mlp'\n","alpha = 5\n","layers = (10,10)\n","\n","targets = ['dG','logcdil_mgml']\n","targets_clean = {\n","    'dG' : 'Delta G',\n","    'logcdil_mgml' : 'Saturation concentration',\n","}\n","\n","print('Models loaded.')"]},{"cell_type":"markdown","metadata":{"id":"rLVOLP3AlHHV"},"source":["# Predict single sequence\n","\n","Paste in an aminoacid sequence to predict phase separation propensities. The termini can be charged (unmodified) or not charged (capped), which should only have minor effects on the prediction if the sequence is long."]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","id":"0F7WtrIChg-F"},"outputs":[],"source":["#@title <b>Run single IDR prediction</font></b>\n","\n","try:\n","    os.rmdir('sample_data')\n","except:\n","    pass\n","\n","NAME = \"LAF1_mod\" #@param {type:\"string\"}\n","SEQUENCE = \"MESNQSNNGGSGNAALNRGGRYVPPHLRGGDGGAAAAASAGGDDRRGGAGGGGYRRGGGNSGGGGGGGYDRGYNDNRDDRDNRGGSGGYGRDRNYEDRGYNGGGGGGGNRGYNNNRGGGGGGYNRQDRGDGGSSNFSRGGYNNRDEGSDNRGSGRSYNNDRRDNGGDGLEHHHHHH\" #@param {type:\"string\"}\n","CHARGE_TERMINI = True # @param {type:'boolean'}\n","TEMPERATURE = \"293 K (fixed)\" # @param ['293 K (fixed)']\n","IONIC_STRENGTH = \"150 mM (fixed)\" # @param ['150 mM (fixed)']\n","\n","seq = SEQUENCE[:]\n","if \" \" in seq:\n","    seq = ''.join(seq.split())\n","    print('Blank character(s) found in the provided sequence. Sequence has been corrected, but check for integrity.')\n","\n","print('='*80)\n","print(f'NAME: {NAME}')\n","print(f'SEQUENCE: {seq}')\n","\n","seqfeats = sequence.SeqFeatures(seq,residues=residues,charge_termini=CHARGE_TERMINI)\n","X = X_from_seq(seq,features,residues=residues,charge_termini=CHARGE_TERMINI,nu_file=nu_file)\n","\n","for target in targets:\n","  print('-'*80)\n","  ys = models[target].predict(X)#,models)\n","  ys_m = np.mean(ys)\n","\n","  if target == 'dG':\n","    output = ys_m\n","    unit = 'kT'\n","    lower = ys_m - 1\n","    upper = ys_m + 1\n","  elif target == 'logcdil_mgml':\n","    output = np.exp(ys_m)\n","    lower = np.exp(ys_m-0.82)\n","    upper = np.exp(ys_m+0.82)\n","    unit = 'mg/mL'\n","\n","  print(f'{targets_clean[target]:25s} = {output:5.1f} {unit:6s} ({lower:.1f} -- {upper:.1f} {unit})')\n","  if target == 'logcdil_mgml':\n","    output_uM = output / seqfeats.mw * 1e6\n","    lower_uM = lower / seqfeats.mw * 1e6\n","    upper_uM = upper / seqfeats.mw * 1e6\n","    print(f'{\"\":25s} = {output_uM:5.1f} {\"uM\":6s} ({lower_uM:.1f} -- {upper_uM:.1f} {\"uM\"})')\n","print('='*80)\n","\n","fig, ax = plt.subplots(figsize=(4,3))\n","\n","target = 'dG'\n","\n","ys = models[target].predict(X)\n","ys_m = np.mean(ys)\n","\n","ax.hist(df_full['dG_pred'],bins=100,histtype='step',density=True,color='black',\n","        label='IDRome')\n","ax.hist(df_full['dG_pred'],bins=100,alpha=0.2,density=True,color='black')\n","\n","ax.axvline(ys_m, color=colors[2], lw=2,label=NAME)\n","\n","ax.set(yscale='log')\n","ax.grid(alpha=0.3)\n","ax.set(xlabel=f'Predicted {targets_clean[target]} [kT]',ylabel='PDF')\n","ax.set(xlim=(None,0.4))\n","ax.legend()\n","fig.tight_layout()\n","fig.savefig(f'dG_distribution_{NAME}.pdf')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"64T7fXW0nQ22","cellView":"form"},"outputs":[],"source":["#@title <b>Relate to input features</font></b>\n","\n","limits = {\n","    'scd' : [-5,6],\n","    # 'scd' : [-8,6],\n","    'ncpr' : [-0.25, 0.25],\n","    'kappa' : [0.0, 0.8],\n","    'faro' : [0., 0.15],\n","    'nu' : [0.2, 0.7],\n","    'nu_svr' : [0.45, 0.62],\n","    'mw' : [0., 60000],\n","    'fcr' : [0., 0.6],\n","    'mean_lambda' : [0.3, 0.6],\n","    'shd' : [1.5, 6.],\n","    'N' : [0, 800],\n","    'ah_ij' : [-0.9,-0.3],\n","}\n","\n","def bin_data(xs,ys,nbins,drange=None):\n","    \"\"\" bin data ys in xs bins, based on numpy.histogram_bin_edges \"\"\"\n","    if drange == None:\n","        xmin, xmax = np.min(xs), np.max(xs)\n","    else:\n","        xmin, xmax = drange[0], drange[1]\n","    bins = np.linspace(xmin,xmax,nbins+1)\n","    y_binned = [[] for _ in range(nbins)]\n","    for x, y in zip(xs,ys):\n","        if x <= bins[0]:\n","            y_binned[0].append(y)\n","        elif x >= bins[-1]:\n","            y_binned[nbins-1].append(y)\n","        else:\n","            for idx in range(nbins):\n","                if x >= bins[idx] and x < bins[idx+1]:\n","                    y_binned[idx].append(y)\n","    return bins, y_binned\n","\n","target = 'dG'\n","\n","ys = models[target].predict(X)\n","dG_single = np.mean(ys)\n","\n","seq = SEQUENCE[:]\n","seqfeats = sequence.SeqFeatures(seq,residues=residues,\n","                                charge_termini=CHARGE_TERMINI,\n","                                nu_file=nu_file)\n","\n","nbins = 84\n","\n","Nfeat = len(features)\n","\n","fig, ax = plt.subplots(2,4,figsize=(10,3.5),sharey=True)\n","\n","for idx, feat in enumerate(features):\n","    axij = ax[idx//4,idx%4]\n","\n","    xs = df_full[feat].to_numpy()\n","    ys = df_full['dG_pred'].to_numpy()\n","    if feat in limits:\n","        drange = limits[feat]\n","    else:\n","        drange = None\n","    edges, y_binned = bin_data(xs,ys,nbins,drange=drange)\n","    edges = (edges[:-1] + edges[1:]) / 2.\n","\n","    y_mean = np.array([np.mean(y) for y in y_binned])\n","    y_std = np.array([np.std(y) for y in y_binned])\n","    yLs = np.array([len(yb) for yb in y_binned])\n","\n","    for e, ym, yst, yL in zip(edges,y_mean,y_std,yLs):\n","        color = colors[0]\n","        markers, caps, bars = axij.errorbar(e,ym,yerr=yst,capsize=1,marker='.',color=color,zorder=10)\n","        [bar.set_alpha(0.3) for bar in bars]\n","        [cap.set_alpha(0.3) for cap in caps]\n","\n","    ax2 = axij.twinx()\n","    ax2.plot(edges,yLs,lw=0.8,alpha=1.,color=colors[1])#'black')\n","    ax2.fill_between(edges,yLs,alpha=0.3,color=colors[1])#'black')\n","    if idx in [3,7]:\n","        ax2.set_ylabel('IDR Frequency',color=colors[1],labelpad=4.0)\n","    if feat == 'faro':\n","        ax2.set(ylim=(0,max(yLs)*2))\n","    elif feat in ['scd', 'ncpr']:\n","        ax2.set(ylim=(0,max(yLs)*2.5))\n","    else:\n","        ax2.set(ylim=(0,max(yLs)*3))\n","    ax2.set_yticks([])\n","    ax2.set_yticklabels([])\n","    ax2.tick_params(axis='y', colors=colors[1])\n","    ax2.grid(False)\n","\n","    x_single = getattr(seqfeats,feat)\n","    axij.plot(x_single,dG_single,'o',color=colors[2],label=NAME)\n","    axij.axvline(x_single,color=colors[2],label=NAME)\n","\n","    axij.set(xlabel=features_clean[feat])\n","    axij.set(ylim=(-7,None))\n","    axij.tick_params(axis='y', colors=colors[0])\n","    axij.grid(ls='dotted',alpha=0.5)\n","\n","# ax[0,0].legend()\n","\n","for idx in range(2):\n","    ax[idx,0].set_ylabel(f'{targets_clean[target]}',color=colors[0])\n","fig.tight_layout(pad=0,w_pad=1.08,h_pad=1.08)#\n","fig.savefig(f'dG_vs_features_{NAME}.pdf')"]},{"cell_type":"markdown","metadata":{"id":"p5mLRrURgona"},"source":["## Predict variant effects\n","\n","Below we substitute each of the residues in the above input sequence one by one with any of the 19 other aminoacids. The resulting plot shows the effect on phase separation propensities. Negative values (red) indicate stronger phase separation, positive values (blue) indicate weaker phase separation compared to the input sequence."]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","id":"EvAjiCoYgqoJ"},"outputs":[],"source":["#@title <b>Run variant effect prediction</font></b>\n","\n","p = np.argsort(residues['lambdas'].to_numpy())\n","sorted_names = residues.index[p]\n","aminoacids_sorted = \"\".join(sorted_names)\n","\n","ah_intgrl_map = sequence.make_ah_intgrl_map(residues)\n","lambda_map = sequence.make_lambda_map(residues)\n","\n","target = 'dG'\n","model = models[target]\n","\n","seq = SEQUENCE[:]\n","wt_idx = np.zeros((len(seq)))\n","dG_map = np.zeros((len(aminoacids_sorted),len(seq)))\n","\n","seq_feats = sequence.SeqFeatures(seq,residues=residues,charge_termini=CHARGE_TERMINI,nu_file=nu_file,\n","            ah_intgrl_map=ah_intgrl_map,lambda_map=lambda_map)\n","\n","X = X_from_seq(seq,features,seq_feats=seq_feats)\n","ys = model.predict(X)\n","dG_wt = np.mean(ys)\n","\n","for idx, s in tqdm(enumerate(seq),total=len(seq)):\n","    wt_idx[idx] = aminoacids_sorted.index(s)\n","    for jdx, a in enumerate(aminoacids_sorted):\n","        seq_new = list(seq)\n","        seq_new[idx] = a\n","        seq_new = \"\".join(seq_new)\n","\n","        seq_feats = sequence.SeqFeatures(seq_new,residues=residues,charge_termini=CHARGE_TERMINI,nu_file=nu_file,\n","                    ah_intgrl_map=ah_intgrl_map,lambda_map=lambda_map)\n","        X = X_from_seq(seq_new,features,seq_feats=seq_feats)\n","        ys = model.predict(X)\n","        ys_m = np.mean(ys)\n","\n","        dG_map[jdx,idx] = ys_m"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","id":"j4MeZOxxiIdi"},"outputs":[],"source":["#@title <b>Plot variant effect results</font></b>\n","\n","from mpl_toolkits.axes_grid1 import make_axes_locatable\n","\n","fig, ax = plt.subplots(figsize=(14,9))#,sharex=True)\n","\n","_ = ax.imshow(dG_map - dG_wt,cmap=plt.cm.RdBu,vmin=-1,vmax=1)\n","divider = make_axes_locatable(ax)\n","cax = divider.append_axes(\"right\", size=\"1%\", pad=0.1)\n","plt.colorbar(_, cax=cax,label=f'$\\Delta \\Delta G$ [kT]')\n","\n","print(f'{targets_clean[\"dG\"]} wt: {dG_wt:.1f} kT')\n","\n","ax.plot(wt_idx,'o',fillstyle='none',color='black',markersize=4)\n","ax.set_xticks(np.arange(len(seq)))\n","ax.set_xticklabels(seq,fontsize=7)\n","ax.set_yticks(np.arange(len(aminoacids_sorted)))\n","ax.set_yticklabels(aminoacids_sorted,fontsize=7)\n","ax.grid(False)\n","ax.set(xlabel='Residues')\n","\n","fig.tight_layout()\n","fig.savefig(f'{NAME}_variants.pdf')"]},{"cell_type":"markdown","metadata":{"id":"_PrG7GsAkt_6"},"source":["# Batch prediction\n","\n","An input fasta file can be uploaded. The code below predicts phase separation propensity and saturation concentrations for all sequences in the fasta file."]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","id":"L9RekX9rtunn"},"outputs":[],"source":["#@title <b>Run batch prediction</b>\n","\n","#@markdown File name\n","FASTA_FILE = \"example.fasta\" #@param {type:\"string\"}\n","CHARGE_TERMINI = True # @param {type:'boolean'}\n","TEMPERATURE = \"293 K (fixed)\" # @param ['293 K (fixed)']\n","IONIC_STRENGTH = \"150 uM (fixed)\" # @param ['150 uM (fixed)']\n","\n","if not os.path.isfile(FASTA_FILE):\n","  print(f'Please upload file {FASTA_FILE}')\n","  uploaded = files.upload()\n","  if FASTA_FILE not in uploaded.keys():\n","    raise NameError(f'Could not find file {FASTA_FILE}')\n","\n","records = sequence.read_fasta(FASTA_FILE)\n","\n","print('-'*80)\n","print(f'FASTA FILE: {FASTA_FILE}')\n","print(f'NUMBER OF SEQUENCES: {len(records)}')\n","\n","#@title <b>Run batch prediction</font></b>\n","\n","df_records = pd.DataFrame(dtype=object)\n","\n","for name, record in tqdm(records.items(),total=len(records)):\n","  seq = str(record.seq)\n","  df_records.loc[name,'Sequence'] = seq\n","  seqfeats = sequence.SeqFeatures(seq,residues=residues,\n","                                  charge_termini=CHARGE_TERMINI,nu_file=nu_file)\n","  for feat in features:\n","    df_records.loc[name,feat] = getattr(seqfeats,feat)\n","  X = X_from_seq(seq,features,residues=residues,\n","                 charge_termini=CHARGE_TERMINI,nu_file=nu_file)\n","  for target in targets:\n","    ys = models[target].predict(X)#,models)\n","    ys_m = np.mean(ys)\n","    if target == 'dG':\n","      df_records.loc[name,'Delta G [kT]'] = ys_m\n","    if target == 'logcdil_mgml':\n","      cdil_mgml = np.exp(ys_m)\n","      df_records.loc[name,'Saturation concentration [mg/mL]'] = cdil_mgml\n","      cdil_uM = cdil_mgml / seqfeats.mw * 1e6\n","      df_records.loc[name,'Saturation concentration [uM]'] = cdil_uM\n","\n","df_records.index.name = 'Name'\n","df_records.to_csv('df_PSprediction.csv')\n","\n","print('\\n')\n","print('='*114)\n","print(f'{\"Name\":20s} {\"Sequence\":33s} {\"Delta G\":>10s} {\"Saturation\":>16s} {\"Saturation\":>16s}')\n","print(f'{\"\":20s} {\"\":33s} {\"\":>10s} {\"concentration\":>16s} {\"concentration\":>16s}')\n","print(f'{\"\":20s} {\"\":33s} {\"[kT]\":>10s} {\"[mg/mL]\":>16s} {\"[uM]\":>16s}')\n","\n","print('='*114)\n","for key, val in df_records.iterrows():\n","  if len(list(val[\"Sequence\"])) > 30:\n","    seqpr = f'{val[\"Sequence\"][:30]:30s}...'\n","  else:\n","    seqpr = f'{val[\"Sequence\"][:30]:30s}'\n","  print(f'{key:20s} {seqpr:33s} {val[\"Delta G [kT]\"]:10.1f} {val[\"Saturation concentration [mg/mL]\"]:16.1f} {val[\"Saturation concentration [uM]\"]:16.1f}')"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","id":"bNX8tEIW29um"},"outputs":[],"source":["#@title <b>Analysis</font></b>\n","\n","fcolor = plt.cm.summer\n","\n","fig, ax = plt.subplots(1,2,figsize=(9,4))\n","\n","for idx, target in enumerate(['Delta G [kT]','Saturation concentration [mg/mL]']):\n","  axij = ax[idx]\n","  axij.hist(df_records[target],bins=20,color=fcolor(0))\n","\n","  axij.set_xlabel(f'{target}')\n","  axij.set_ylabel('Counts')\n","  axij.grid(alpha=0.3)\n","fig.tight_layout()"]},{"cell_type":"code","execution_count":null,"metadata":{"cellView":"form","id":"-SCC7r8ltUTy"},"outputs":[],"source":["#@title <b>Download dataframe</font></b>\n","\n","files.download('df_PSprediction.csv')"]}],"metadata":{"colab":{"provenance":[{"file_id":"https://github.com/KULL-Centre/_2024_buelow_PSpred/blob/twin2pipsa/PSLab_TWIN2PIPSA.ipynb","timestamp":1741602787081},{"file_id":"https://github.com/KULL-Centre/_2024_buelow_PSpred/blob/main/PSLab.ipynb","timestamp":1740476816461}]},"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.15"}},"nbformat":4,"nbformat_minor":0}